File: libs/go/resilience/retry/retry.go
// Package retry provides generic retry functionality with configurable backoff.
package retry

import (
	"context"
	"errors"
	"fmt"
	"math"
	"math/rand"
	"time"
)

// Config holds retry configuration.
type Config struct {
	MaxAttempts int
	BaseDelay   time.Duration
	MaxDelay    time.Duration
	Multiplier  float64
	Jitter      float64
	RetryIf     func(error) bool
}

// DefaultConfig returns a default configuration.
func DefaultConfig() Config {
	return Config{
		MaxAttempts: 3,
		BaseDelay:   100 * time.Millisecond,
		MaxDelay:    10 * time.Second,
		Multiplier:  2.0,
		Jitter:      0.1,
		RetryIf:     func(error) bool { return true },
	}
}

// Option configures retry behavior.
type Option func(*Config)

// WithMaxAttempts sets the maximum number of retry attempts.
func WithMaxAttempts(n int) Option {
	return func(c *Config) { c.MaxAttempts = n }
}

// WithBaseDelay sets the base delay between retries.
func WithBaseDelay(d time.Duration) Option {
	return func(c *Config) { c.BaseDelay = d }
}

// WithMaxDelay sets the maximum delay between retries.
func WithMaxDelay(d time.Duration) Option {
	return func(c *Config) { c.MaxDelay = d }
}

// WithMultiplier sets the exponential backoff multiplier.
func WithMultiplier(m float64) Option {
	return func(c *Config) { c.Multiplier = m }
}

// WithJitter sets the jitter percentage (0-1).
func WithJitter(percent float64) Option {
	return func(c *Config) { c.Jitter = percent }
}

// WithRetryIf sets the retry predicate.
func WithRetryIf(predicate func(error) bool) Option {
	return func(c *Config) { c.RetryIf = predicate }
}

// RetryError wraps the last error with attempt information.
type RetryError struct {
	Attempts int
	LastErr  error
}

func (e *RetryError) Error() string {
	return fmt.Sprintf("all %d retry attempts exhausted: %v", e.Attempts, e.LastErr)
}

func (e *RetryError) Unwrap() error {
	return e.LastErr
}

// Retry executes the operation with retry policy.
func Retry[T any](ctx context.Context, op func() (T, error), opts ...Option) (T, error) {
	config := DefaultConfig()
	for _, opt := range opts {
		opt(&config)
	}

	var zero T
	var lastErr error

	for attempt := 1; attempt <= config.MaxAttempts; attempt++ {
		result, err := op()
		if err == nil {
			return result, nil
		}

		lastErr = err

		// Check if we should retry
		if !config.RetryIf(err) {
			return zero, err
		}

		// Check if this was the last attempt
		if attempt >= config.MaxAttempts {
			break
		}

		// Calculate delay with exponential backoff
		delay := calculateDelay(attempt, config)

		// Wait or check for context cancellation
		select {
		case <-ctx.Done():
			return zero, ctx.Err()
		case <-time.After(delay):
		}
	}

	return zero, &RetryError{
		Attempts: config.MaxAttempts,
		LastErr:  lastErr,
	}
}

// RetryWithContext executes the operation with retry policy and context-aware operation.
func RetryWithContext[T any](ctx context.Context, op func(context.Context) (T, error), opts ...Option) (T, error) {
	return Retry(ctx, func() (T, error) {
		return op(ctx)
	}, opts...)
}

func calculateDelay(attempt int, config Config) time.Duration {
	// Exponential backoff
	delay := float64(config.BaseDelay) * math.Pow(config.Multiplier, float64(attempt-1))

	// Cap at max delay
	if delay > float64(config.MaxDelay) {
		delay = float64(config.MaxDelay)
	}

	// Add jitter
	if config.Jitter > 0 {
		jitter := delay * config.Jitter * (rand.Float64()*2 - 1)
		delay += jitter
	}

	return time.Duration(delay)
}

// IsRetryError checks if the error is a RetryError.
func IsRetryError(err error) bool {
	var retryErr *RetryError
	return errors.As(err, &retryErr)
}

// GetRetryAttempts returns the number of attempts from a RetryError.
func GetRetryAttempts(err error) int {
	var retryErr *RetryError
	if errors.As(err, &retryErr) {
		return retryErr.Attempts
	}
	return 0
}

// Do is a simpler version of Retry for operations that don't return a value.
func Do(ctx context.Context, op func() error, opts ...Option) error {
	_, err := Retry(ctx, func() (struct{}, error) {
		return struct{}{}, op()
	}, opts...)
	return err
}

File: libs/rust/rust-common/src/retry.rs
//! Retry policy implementation with exponential backoff.
//!
//! This module provides a configurable retry mechanism for handling
//! transient failures in distributed systems.

use std::time::Duration;
use crate::PlatformError;

/// Retry policy configuration.
#[derive(Debug, Clone)]
pub struct RetryConfig {
    /// Maximum number of retry attempts
    pub max_retries: u32,
    /// Initial delay between retries
    pub initial_delay: Duration,
    /// Maximum delay between retries
    pub max_delay: Duration,
    /// Multiplier for exponential backoff
    pub multiplier: f64,
    /// Whether to add jitter to delays
    pub jitter: bool,
}

impl Default for RetryConfig {
    fn default() -> Self {
        Self {
            max_retries: 3,
            initial_delay: Duration::from_millis(100),
            max_delay: Duration::from_secs(10),
            multiplier: 2.0,
            jitter: true,
        }
    }
}

impl RetryConfig {
    /// Create a new retry config with custom max retries.
    #[must_use]
    pub const fn with_max_retries(mut self, max_retries: u32) -> Self {
        self.max_retries = max_retries;
        self
    }

    /// Create a new retry config with custom initial delay.
    #[must_use]
    pub const fn with_initial_delay(mut self, delay: Duration) -> Self {
        self.initial_delay = delay;
        self
    }

    /// Create a new retry config with custom max delay.
    #[must_use]
    pub const fn with_max_delay(mut self, delay: Duration) -> Self {
        self.max_delay = delay;
        self
    }

    /// Create a new retry config without jitter.
    #[must_use]
    pub const fn without_jitter(mut self) -> Self {
        self.jitter = false;
        self
    }
}

/// Retry policy for executing operations with automatic retries.
#[derive(Debug, Clone)]
pub struct RetryPolicy {
    config: RetryConfig,
}

impl RetryPolicy {
    /// Create a new retry policy with the given configuration.
    #[must_use]
    pub const fn new(config: RetryConfig) -> Self {
        Self { config }
    }

    /// Create a retry policy with default configuration.
    #[must_use]
    pub fn with_defaults() -> Self {
        Self::new(RetryConfig::default())
    }

    /// Calculate the delay for a given attempt number.
    ///
    /// Uses exponential backoff with optional jitter.
    #[must_use]
    pub fn delay_for_attempt(&self, attempt: u32) -> Duration {
        let base_delay = self.config.initial_delay.as_millis() as f64
            * self.config.multiplier.powi(attempt as i32);

        let delay_ms = base_delay.min(self.config.max_delay.as_millis() as f64);

        let final_delay = if self.config.jitter {
            // Add up to 25% jitter
            let jitter_factor = 1.0 + (rand::random::<f64>() * 0.25);
            delay_ms * jitter_factor
        } else {
            delay_ms
        };

        Duration::from_millis(final_delay as u64)
    }

    /// Check if an error should be retried.
    #[must_use]
    pub fn should_retry(&self, error: &PlatformError, attempt: u32) -> bool {
        attempt < self.config.max_retries && error.is_retryable()
    }

    /// Execute an async operation with retries.
    ///
    /// # Errors
    ///
    /// Returns the last error if all retries are exhausted.
    pub async fn execute<F, Fut, T>(&self, mut operation: F) -> Result<T, PlatformError>
    where
        F: FnMut() -> Fut,
        Fut: std::future::Future<Output = Result<T, PlatformError>>,
    {
        let mut attempt = 0;
        loop {
            match operation().await {
                Ok(result) => return Ok(result),
                Err(error) => {
                    if !self.should_retry(&error, attempt) {
                        return Err(error);
                    }
                    let delay = self.delay_for_attempt(attempt);
                    tokio::time::sleep(delay).await;
                    attempt += 1;
                }
            }
        }
    }

    /// Get the maximum number of retries.
    #[must_use]
    pub const fn max_retries(&self) -> u32 {
        self.config.max_retries
    }
}

impl Default for RetryPolicy {
    fn default() -> Self {
        Self::with_defaults()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_default_config() {
        let config = RetryConfig::default();
        assert_eq!(config.max_retries, 3);
        assert_eq!(config.initial_delay, Duration::from_millis(100));
    }

    #[test]
    fn test_delay_calculation_no_jitter() {
        let config = RetryConfig::default().without_jitter();
        let policy = RetryPolicy::new(config);

        let delay0 = policy.delay_for_attempt(0);
        let delay1 = policy.delay_for_attempt(1);
        let delay2 = policy.delay_for_attempt(2);

        assert_eq!(delay0, Duration::from_millis(100));
        assert_eq!(delay1, Duration::from_millis(200));
        assert_eq!(delay2, Duration::from_millis(400));
    }

    #[test]
    fn test_max_delay_cap() {
        let config = RetryConfig::default()
            .without_jitter()
            .with_max_delay(Duration::from_millis(150));
        let policy = RetryPolicy::new(config);

        let delay2 = policy.delay_for_attempt(2);
        assert_eq!(delay2, Duration::from_millis(150));
    }

    #[test]
    fn test_should_retry() {
        let policy = RetryPolicy::with_defaults();

        // Retryable error within limit
        assert!(policy.should_retry(&PlatformError::RateLimited, 0));
        assert!(policy.should_retry(&PlatformError::RateLimited, 2));

        // Retryable error at limit
        assert!(!policy.should_retry(&PlatformError::RateLimited, 3));

        // Non-retryable error
        assert!(!policy.should_retry(&PlatformError::NotFound("test".to_string()), 0));
    }

    #[tokio::test]
    async fn test_execute_success() {
        let policy = RetryPolicy::with_defaults();
        let result: Result<i32, PlatformError> = policy.execute(|| async { Ok(42) }).await;
        assert_eq!(result.unwrap(), 42);
    }

    #[tokio::test]
    async fn test_execute_non_retryable_error() {
        let policy = RetryPolicy::with_defaults();
        let result: Result<i32, PlatformError> = policy
            .execute(|| async { Err(PlatformError::NotFound("test".to_string())) })
            .await;
        assert!(result.is_err());
    }
}
